python_modules/dagster-graphql/dagster_graphql/implementation/execution/backfill.py

```diff
@@ -318,3 +324,76 @@ def resume_partition_backfill(
 
     graphene_info.context.instance.update_backfill(backfill.with_status(BulkActionStatus.REQUESTED))
     return GrapheneResumeBackfillSuccess(backfill_id=backfill_id)
+
+
+def retry_partition_backfill(
+    graphene_info: "ResolveInfo", backfill_id: str
+) -> "GrapheneRetryBackfillSuccess":
+    from ...schema.backfill import GrapheneRetryBackfillSuccess
+
+    backfill = graphene_info.context.instance.get_backfill(backfill_id)
+    if not backfill:
+        check.failed(f"No backfill found for id: {backfill_id}")
+
+    if backfill.status not in BULK_ACTION_COMPLETED_STATUSES:
+        raise DagsterInvariantViolationError(
+            f"Cannot retry backfill {backfill_id} because it is still in progress."
+        )
+
+    if backfill.is_asset_backfill:
+        asset_backfill_data = backfill.get_asset_backfill_data(graphene_info.context.asset_graph)
+        if (
+            asset_backfill_data.failed_and_downstream_subset.num_partitions_and_non_partitioned_assets
+            == 0
+        ):
+            raise DagsterInvariantViolationError(
+                "Cannot retry an asset backfill that has no failed assets."
+            )
+        asset_graph = graphene_info.context.asset_graph
+        assert_permission_for_asset_graph(
+            graphene_info,
+            asset_graph,
+            list(asset_backfill_data.failed_and_downstream_subset.asset_keys),
+            Permissions.LAUNCH_PARTITION_BACKFILL,
+        )
+
+        new_backfill = PartitionBackfill.from_asset_graph_subset(
+            backfill_id=make_new_backfill_id(),
+            asset_graph_subset=asset_backfill_data.failed_and_downstream_subset,
+            dynamic_partitions_store=graphene_info.context.instance,
+            tags={
+                **backfill.tags,
+                PARENT_RUN_ID_TAG: backfill.backfill_id,
+                ROOT_RUN_ID_TAG: backfill.tags.get(PARENT_RUN_ID_TAG, backfill.backfill_id),
+            },
+            backfill_timestamp=get_current_timestamp(),
+            title=f"Retry of {backfill.title}" if backfill.title else None,
+            description=backfill.description,
+        )
+    else:
+        partition_set_origin = check.not_none(backfill.partition_set_origin)
+        location_name = partition_set_origin.selector.location_name
+        assert_permission_for_location(
+            graphene_info, Permissions.LAUNCH_PARTITION_BACKFILL, location_name
+        )
+
+        new_backfill = PartitionBackfill(
+            backfill_id=make_new_backfill_id(),
+            partition_set_origin=backfill.partition_set_origin,
+            status=BulkActionStatus.REQUESTED,
+            partition_names=backfill.partition_names,
+            from_failure=True,
+            reexecution_steps=backfill.reexecution_steps,
```

Comment:
@sryza I'm less familiar with how job backfills work, and i see that there are different paths taken in the daemon if `fromFailure` is set or `reexecution_steps` is set https://github.com/dagster-io/dagster/blob/47a609d07c76f3b69f5978f97e9d62294d06c43c/python_modules/dagster/dagster/_core/execution/job_backfill.py#L420-L434

i'm looking through what the differences are to decide what the right thing to do here is, but lmk if you have opinions

Reply:
can ignore for now. i need to look at the FE code a bit more. it doesn't look like it allows backfill retries for job backfills (even though the launch mutation would accept a parameter to retry* the backfill) 

* it's not exactly a retry, it looks like setting `fromFailure` means that before launching a run for a partition, the daemon will only launch the run if the most recent run for that partitioned failed. which is kind of the same as retrying the backfill, but allows for ad-hoc runs of jobs to influence if a run is submitted or not